pkg/config/storage.go

```diff
@@ -0,0 +1,76 @@
+package config
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const driverUnset = ""
+const driverNone = "none"
+const driverLVMS = "lvms"
+
+const csiComponentNone = "none"
+const csiComponentSnapshot = "snapshot-controller"
+const csiComponentSnapshotWebhook = "snapshot-webhook"
+
+// Storage represents a subfield of the microshift config data structure. Its purpose to provide a user
+// facing interface to control whether MicroShift should deploy LVMS on startup.
+type Storage struct {
+	// One of ["lvms", "none", ""]
+	Driver        string         `json:"driver,omitempty"`
+	CSIComponents []string `json:"with-csi-components,omitempty"`
+}
+
+func (s Storage) driverIsValid() (isSupported bool) {
+	return sets.New[string](driverNone, driverLVMS, driverUnset).Has(s.Driver)
+}
+
+func (s Storage) csiComponentsAreValid() ([]string, bool) {
+	supported := sets.New[string](csiComponentNone, csiComponentSnapshot, csiComponentSnapshotWebhook)
+	unsupported := sets.New[string]()
+
+	for _, cfgComp := range s.CSIComponents {
+		if !supported.Has(cfgComp) {
+			unsupported.Insert(cfgComp)
+		}
+	}
+	return unsupported.UnsortedList(), unsupported.Len() == 0
+}
+
+func (s Storage) noneIsMutuallyExclusive() ([]string, bool) {
+	components := sets.New[string](s.CSIComponents...)
+	if components.Has(csiComponentNone) && components.Len() > 1 {
+		components.Delete(csiComponentNone)
+		ret := make([]string, 0)
+		for v := range components {
+			ret = append(ret, v)
+		}
+		return ret, false
```

Comment:
https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set.UnsortedList
```suggestion
		return components.UnsortedList(), false
```

Reply:
There was a good reason I didn't initially use this, and I cannot remember at all what it was. I'll retest with the UnsortedList and if it doesn't become apparent, I'll commit this change.